lib/parsers/ts-utils.js

```diff
@@ -0,0 +1,63 @@
+const fs = require('node:fs');
+const ts = require('typescript');
+const { transformForLint } = require('./transform');
+
+const TsProgramMap = {};
+const ProcessedFiles = new Set();
+
+module.exports.getTsProgram = function getTsProgram(projectFile, filePath, code) {
+  // we need to recreate the program if a file is processed multiple times
+  // this can happen during tests and when eslint is fixing the files
+  // during fixing, eslint can do multiple cycles
+  if (!TsProgramMap[projectFile] || ProcessedFiles.has(filePath)) {
+    ProcessedFiles.add(filePath);
+    const config = ts.getParsedCommandLineOfConfigFile(
+      projectFile,
+      {},
+      {
+        ...ts.sys,
+        readDirectory(dir) {
+          const results = ts.sys.readDirectory(dir);
+          return results.map((f) => f.replace(/\.gts$/, '.ts'));
+        },
+      }
+    );
+
+    const host = ts.createCompilerHost(config.options);
+    host.fileExists = function (fileName) {
+      return fs.existsSync(fileName.replace(/\.ts$/, '.gts')) || fs.existsSync(fileName);
+    };
+
+    const readDirectory = host.readDirectory;
+    host.readDirectory = function (dir) {
+      const results = readDirectory.call(this, dir);
+      return results.map((f) => f.replace(/\.gts$/, '.ts'));
+    };
+    host.readFile = function (fname) {
+      if (fname === filePath.replaceAll('\\', '/')) {
+        return code;
+      }
+      let fileName = fname;
+      let content = '';
+      try {
+        content = fs.readFileSync(fileName).toString();
+      } catch {
+        fileName = fileName.replace(/\.ts$/, '.gts');
+        content = fs.readFileSync(fileName).toString();
+      }
+      if (fileName.endsWith('.gts')) {
+        content = transformForLint(content).output;
+      }
+      content = content.replaceAll(/\.gts(["'])/g, '.ts$1 ');
+      return content;
+    };
+
+    TsProgramMap[projectFile] = ts.createProgram({
```

Comment:
@patricklx This PR is looking pretty nice so far, great work on this.
Does it continue to work when we are doing file changes in our IDE?
I think we will need to use `ts.createWatchProgram` in cases when this is running as part of eslint extension for vs-code or similar?